src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```diff
@@ -0,0 +1,748 @@
+        std::vector<hcn::CreateEndpointParameters> params{};
+
+        // The primary endpoint (management)
+        hcn::CreateEndpointParameters primary_endpoint{};
+        primary_endpoint.network_guid = primary_network_guid;
+        primary_endpoint.endpoint_guid = mac2uuid(description.default_mac_address);
+        primary_endpoint.mac_address = description.default_mac_address;
+        replace_colon_with_dash(primary_endpoint.mac_address.value());
+        params.push_back(primary_endpoint);
+
+        // Additional endpoints, a.k.a. extra interfaces.
+        for (const auto& v : description.extra_interfaces)
+        {
+            hcn::CreateEndpointParameters extra_endpoint{};
+            extra_endpoint.network_guid = multipass::utils::make_uuid(v.id).toStdString();
+            extra_endpoint.endpoint_guid = mac2uuid(v.mac_address);
+            extra_endpoint.mac_address = v.mac_address;
+            replace_colon_with_dash(extra_endpoint.mac_address.value());
+            params.push_back(extra_endpoint);
+        }
+        return params;
+    }();
+
+    for (const auto& endpoint : create_endpoint_params)
+    {
+        // There might be remnants from an old VM, remove the endpoint if exist before
+        // creating it again.
+        if (HCN().delete_endpoint(endpoint.endpoint_guid))
+        {
+            mpl::debug(log_category,
+                       "The endpoint {} was already present for the VM {}, removed it.",
+                       endpoint.endpoint_guid,
+                       get_name());
+        }
+        if (const auto& [status, msg] = HCN().create_endpoint(endpoint); !status)
+        {
+            throw CreateEndpointException{"create_endpoint failed with {}", status};
+        }
+    }
+
+    // E_INVALIDARG means there's no such VM.
+    // Create the VM from scratch.
+    const auto create_compute_system_params = [this, &create_endpoint_params]() {
+        hcs::CreateComputeSystemParameters params{};
+        params.name = description.vm_name;
+        params.memory_size_mb = description.mem_size.in_megabytes();
+        params.processor_count = description.num_cores;
+
+        hcs::HcsScsiDevice primary_disk{hcs::HcsScsiDeviceType::VirtualDisk()};
+        primary_disk.name = "Primary disk";
+        primary_disk.path = get_primary_disk_path();
+        primary_disk.read_only = false;
+        params.scsi_devices.push_back(primary_disk);
+
+        hcs::HcsScsiDevice cloudinit_iso{hcs::HcsScsiDeviceType::Iso()};
+        cloudinit_iso.name = "cloud-init ISO file";
+        cloudinit_iso.path = description.cloud_init_iso.toStdString();
+        cloudinit_iso.read_only = true;
+        params.scsi_devices.push_back(cloudinit_iso);
+
+        const auto create_endpoint_to_network_adapter = [this](const auto& create_params) {
+            hcs::HcsNetworkAdapter network_adapter{};
+            network_adapter.endpoint_guid = create_params.endpoint_guid;
+            if (!create_params.mac_address)
```

Comment:
When could this happen (other than by a logical error in the program)?

Reply:
It's optional for a network adapter to have a static mac addr set, so the code is checking whether the optional has a value or not before deref'ing it.

I feel like I'm being punished for practicing defensive programming. Does the code has to assume the caller or the systems are behaving as expected at all times? It's just a piece of code validating its assumptions and I fail to see what's bad about it.